build-logic/java-api-extractor/src/main/java/org/gradle/internal/tools/api/impl/JavaApiMemberWriter.java

```diff
@@ -80,6 +83,17 @@ public void writeMethod(MethodMember method) {
             method.getExceptions().toArray(new String[0]));
         writeMethodAnnotations(mv, method.getAnnotations());
         writeMethodAnnotations(mv, method.getParameterAnnotations());
+
+        // Without this parameter annotations for non-static inner class constructors would
+        // end up having one more parameter than they should, because the first parameter
+        // (pointing to `this` of the enclosing type) should not be listed.
+        // See https://gitlab.ow2.org/asm/asm/-/issues/318023
+        if (method.getSignature() != null) {
```

Comment:
❓ Should we do that only for constructors or is it ok to do for all methods?

Reply:
That's not actually a problem, because we can always determine the proper list of arguments from the signature.

However, you just made me realize there is another problem. Signatures are only available if there's any generics involved. If I change the example constructor to have no generics, the problem persists. :/

Back to the drawing board I guess.

Reply:
Hmm, I would expect signature is always present

Reply:
It isn't unfortunately (just tested).